### DISCUSSION TITLE
adding commits to earlier branches?

### DISCUSSION THREAD

**Discussion by alan-nf** [arxanas/git-branchless#315] [Q&A] [answered]
Firstly, thanks for making this great project.

Now my question: I have a repo that looks like this:

```
➜  git:(feature4) git sl
◇ 70bcbf7a 1h (main, remote origin/main) Initial commit
┃
◯ 2d27ec27 45m (feature1) feature1 work
┃
◯ bff4f430 45m (feature2) feature2 work
┃
◯ 5cd580b9 2m (feature3) feature3 work
┃
● e106b908 2m (feature4) feature4 work
```

My goal is to make another commit to the `feature2` branch and end up with a repo that looks like this:
```
➜ git:(feature4) git sl
◇ 70bcbf7a 1h (main, remote origin/main) Initial commit
┃
◯ 2d27ec27 48m (feature1) feature1 work
┃
◯ bff4f430 48m feature2 work
┃
◯ 5e3fa576 1m (feature2) more feature2 work     # <=== new commit
┃
◯ 38dadd89 1m (feature3) feature3 work
┃
● 647556fd 1m (feature4) feature4 work
```

If I didn't use branches this would be trivial. I could just make a fixup commit, or a full commit and then `git rebase -i`. But I'm having trouble figuring out what the `git branchless` workflow should be if I'm using branches. This works but is a LOT of steps:

```
# Make the new commit
echo "more stuff" >> feature2.txt
git commit -a -m "more feature2 work"

# New repo state
➜  git:(feature4) git sl
◇ 70bcbf7a 1h (main, remote origin/main) Initial commit
┃
◯ 2d27ec27 46m (feature1) feature1 work
┃
◯ bff4f430 46m (feature2) feature2 work
┃
◯ 5cd580b9 2m (feature3) feature3 work
┃
◯ e106b908 2m feature4 work
┃
● 86e14ef3 1s (feature4) more feature2 work     # <=== new commit (wrong branch, wrong location)

# Move the new commit up using git rebase
➜  git:(feature4) git rebase -i HEAD~5
➜  git:(feature4) git restack

# New repo state
➜ git:(feature4) git sl
◇ 70bcbf7a 1h (main, remote origin/main) Initial commit
┃
◯ 2d27ec27 1h (feature1) feature1 work
┃
◯ bff4f430 1h (feature2) feature2 work
┃
◯ a841d4e9 4s more feature2 work     # <=== new commit (wrong branch, correct location)
┃
◯ 641cf3fa 3s (feature3) feature3 work
┃
● 8d6f9597 3s (feature4) feature4 work

# fix the branch pointer
➜  git:(feature4) git branch --force feature2 a841d4e9

# New repo state
➜  git:(feature4) git sl
◇ 70bcbf7a 1h (main, remote origin/main) Initial commit
┃
◯ 2d27ec27 1h (feature1) feature1 work
┃
◯ bff4f430 1h feature2 work
┃
◯ a841d4e9 1m (feature2) more feature2 work     # <=== new commit (correct branch, correct location)
┃
◯ 641cf3fa 1m (feature3) feature3 work
┃
● 8d6f9597 1m (feature4) feature4 work
```

Any tips on better ways of handling this?

**Comment by alan-nf** [comment]
NOTE: I tried a different way of doing this (make the new commit during an interactive rebase) but that method is equally verbose and also broken (https://github.com/arxanas/git-branchless/issues/316)

**Comment by arxanas** [comment, marked as answer]
EDIT: for the time being, I've added a workflow documentation item titled "Adding a commit to an earlier branch" which points to this answer.

As a general resource, you might want to read this article, as it may answer some workflow questions: https://github.com/arxanas/git-branchless/wiki/Workflow:-Editing-an-old-commit's-contents. I've added a to-do to https://github.com/arxanas/git-branchless/issues/115 for your specific use-case, and hopefully I'll get around to it someday 😅 .

For this case, here's what I would do:

1\. `git co feature-2`
2\. `git commit -am 'more feature2 work'`.

- EDIT: @sebastianst points out that in recent versions of git-branchless, rather than run `git commit`, you can instead run `git record -Im 'more feature2 work'` to insert the commit and skip step 3.

Your smartlog should look like this:

```
◇ 70bcbf7a 1h (main, remote origin/main) Initial commit
┃
◯ 2d27ec27 45m (feature1) feature1 work
┃
◯ bff4f430 45m feature2 work
┣━┓
┃ ◯ 5cd580b9 2m (feature3) feature3 work
┃ ┃
┃ ◯ e106b908 2m (feature4) feature4 work
┃
● 5cd580b9 2m (feature2) more feature2 work
```

3\. `git move -s feature-3` (which means "move the commits starting at `feature3` on top of this commit"; see [the documentation](https://github.com/arxanas/git-branchless/wiki/Command:-git-move))

Now it should look like this:

```
◇ 70bcbf7a 1h (main, remote origin/main) Initial commit
┃
◯ 2d27ec27 45m (feature1) feature1 work
┃
◯ bff4f430 45m feature2 work
┃
● 5cd580b9 2m (feature2) more feature2 work
┃
◯ 5cd580b9 2m (feature3) feature3 work
┃
◯ e106b908 2m (feature4) feature4 work
```

Generally speaking, you don't need to heavily rely on `git rebase -i` to edit previous commits. You can instead just check out the given commit and start making your changes there. Commands like `git move`, `git amend`, and `git restack` will try to be helpful about restacking descendant branches.

**Reply by alan-nf** [reply]
Oh, of course. I'm not sure why I didn't come up with that now that you said it :p

Thanks!

**Reply by claytonrcarter** [reply]
Just a point of clarification, because this also tripped me up when I was getting started w/ `branchless`:

In this case, `git move -s feature3` is correct *only* because branch `feature3` only has 1 commit. If there were more commits on that branch, then you would want to use `git move -b feature3` instead.

**Reply by pokey** [reply]
I believe this could now be done generically using revsets:

```
git move -x 'descendants(parents(.)) - ancestors(.)'
```

This way you could map a command `git commit-insert` to do the commit, followed by the above command

**Reply by arxanas** [reply]
It would also be nice to have a `siblings` revset function, but I haven't gotten around to implementing that.

**Reply by sebastianst** [reply]
Couldn't you also just use `git record --insert` for this workflow?

**Reply by alan-nf** [reply]
@sebastianst that would certainly let you take @arxanas [three-command answer](https://github.com/arxanas/git-branchless/discussions/315#discussioncomment-2445315) and turn it into two commands, yes. It didn't exist at the time of this discussion but it is helpful here.

**Reply by arxanas** [reply]
@sebastianst thanks for pointing out, I've updated the answer to mention `git record --insert`!

**Reply by pokey** [reply]
@arxanas you might want to delete that final parenthetical as well 😊

**Reply by arxanas** [reply]
@pokey yes, thanks!
